### PR TITLE
feat(core): orm

### PR DESCRIPTION
**Pull request by forehalo**

### CODE REVIEW

**Comment by forehalo** [comment]
* **#6536** <a href="https://app.graphite.dev/github/pr/toeverything/AFFiNE/6536?utm_source=stack-comment-icon" target="_blank">Graphite</a> 👈
* `canary`

This stack of pull requests is managed by Graphite. <a href="https://stacking.dev/?utm_source=stack-comment">Learn more about stacking.</a>
<h2></h2>

 Join @forehalo and the rest of your teammates on <a href="https://graphite.dev?utm-source=stack-comment">Graphite <b>Graphite</b></a>

File: packages/common/infra/src/orm/core/table.ts [hunk 1]
```diff
@@ -0,0 +1,156 @@
...
+
+export type UpdateEntityInput<T extends TableSchemaBuilder> = Pretty<{
+  [key in NonPrimaryKeyFields<T>]?: T[key] extends FieldSchemaBuilder<
+    infer Type
+  >
+    ? Type
+    : never;
+}>;
+
+export class Table<T extends TableSchemaBuilder> {
```
Comment (EYHN) [inline packages/common/infra/src/orm/core/table.ts:77]: need list() and list$()

File: packages/common/infra/src/orm/core/adapters/yjs.ts [hunk 2]
```diff
@@ -0,0 +1,112 @@
...
+      return null;
+    }
+
+    return record.toJSON();
+  }
+
+  async subscribe(table: string, key: any, callback: (data: any) => void) {
+    const record: YMap<any> = this.getTable(table).getMap(key.toString());
+
+    callback(this.value(record));
```
Comment (EYHN) [inline packages/common/infra/src/orm/core/adapters/yjs.ts]: subscribe will be called frequently. and `this.value(record)` looks heavy, can we share the same reference in multiple subscribers

Comment (forehalo) [reply packages/common/infra/src/orm/core/adapters/yjs.ts]: sure, the table has a subscriber cache

File: packages/common/infra/src/orm/core/table.ts [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+
+  get(key: PrimaryKeyFieldType<T>) {
+    return this.adapter.read(this.name, key);
+  }
+
+  async get$(key: PrimaryKeyFieldType<T>) {
+    let subject$: Subject<any> | undefined = this.subscribedKeys.get(key);
+
+    if (!subject$) {
+      subject$ = new Subject();
```
Comment (EYHN) [inline packages/common/infra/src/orm/core/table.ts]: i think we should use BehaviorSubject

Comment (forehalo) [reply packages/common/infra/src/orm/core/table.ts]: I found using subject here is a bad idea.
there is no way to destroy the subscription

**Review by Brooooooklyn — APPROVED** [review]

**Review by EYHN — APPROVED** [review]

**Pull request merged**